operation/fhir-operation-cqf/src/main/java/com/ibm/fhir/operation/cqf/CareGapsOperation.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * (C) Copyright IBM Corp. 2021
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package com.ibm.fhir.operation.cqf;
+
+import static com.ibm.fhir.cql.helpers.ModelHelper.fhirstring;
+
+import java.time.ZoneOffset;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import javax.ws.rs.core.MultivaluedHashMap;
+import javax.ws.rs.core.MultivaluedMap;
+
+import org.opencds.cqf.cql.engine.data.DataProvider;
+import org.opencds.cqf.cql.engine.retrieve.RetrieveProvider;
+import org.opencds.cqf.cql.engine.runtime.Interval;
+import org.opencds.cqf.cql.engine.terminology.TerminologyProvider;
+
+import com.ibm.fhir.cql.helpers.DataProviderFactory;
+import com.ibm.fhir.cql.helpers.FhirBundleCursor;
+import com.ibm.fhir.cql.helpers.ParameterMap;
+import com.ibm.fhir.ecqm.common.MeasureReportType;
+import com.ibm.fhir.exception.FHIROperationException;
+import com.ibm.fhir.model.resource.Bundle;
+import com.ibm.fhir.model.resource.Measure;
+import com.ibm.fhir.model.resource.MeasureReport;
+import com.ibm.fhir.model.resource.OperationDefinition;
+import com.ibm.fhir.model.resource.Parameters;
+import com.ibm.fhir.model.resource.Parameters.Parameter;
+import com.ibm.fhir.model.resource.Resource;
+import com.ibm.fhir.model.type.UnsignedInt;
+import com.ibm.fhir.model.type.code.BundleType;
+import com.ibm.fhir.model.type.code.ResourceType;
+import com.ibm.fhir.registry.FHIRRegistry;
+import com.ibm.fhir.search.SearchConstants;
+import com.ibm.fhir.server.operation.spi.FHIROperationContext;
+import com.ibm.fhir.server.operation.spi.FHIRResourceHelpers;
+
+public class CareGapsOperation extends AbstractMeasureOperation {
+
+    public static final String PARAM_IN_TOPIC = "topic";
+    public static final String PARAM_IN_SUBJECT = "subject";
+    public static final String PARAM_OUT_RETURN = "return";
+
+    @Override
+    protected OperationDefinition buildOperationDefinition() {
+        return FHIRRegistry.getInstance().getResource("http://hl7.org/fhir/OperationDefinition/Measure-care-gaps", OperationDefinition.class);
+    }
+
+    @Override
+    protected Parameters doInvoke(FHIROperationContext operationContext, Class<? extends Resource> resourceType, String logicalId, String versionId,
+        Parameters parameters, FHIRResourceHelpers resourceHelper) throws FHIROperationException {
+       
+        ParameterMap paramMap = new ParameterMap(parameters);
+
+        ZoneOffset zoneOffset = getZoneOffset(paramMap);
+        Interval measurementPeriod = getMeasurementPeriod(paramMap,zoneOffset);
+        
+        Parameter pTopic = paramMap.getSingletonParameter(PARAM_IN_TOPIC);
+        String topic = ((com.ibm.fhir.model.type.String)pTopic.getValue()).getValue();
+        
+        Parameter pSubject = paramMap.getSingletonParameter(PARAM_IN_SUBJECT);
+        String subject = ((com.ibm.fhir.model.type.String)pSubject.getValue()).getValue();
```

Comment:
there are a lot of edge cases... data absent... which results in an empty/null string. 
this is probably an area that needs hardening.

Reply:
The ParameterMap.getSingletonParameter method will throw an exception if the required data is not found. This is in addition to the framework-level validation of the required input parameters. There isn't much chance this gets invalid data.